conjure-macros/src/lib.rs

```diff
@@ -0,0 +1,805 @@
+/// where
+///     C: AsyncClient + Sync + Send,
+///     C::ResponseBody: 'static + Send,
+/// {
+///     let client = MyServiceAsyncClient::new(client);
+///     client.create_yak(auth, None, "my cool yak").await?;
+///
+///     Ok(())
+/// }
+/// ```
+#[proc_macro_attribute]
+pub fn conjure_client(
+    _attr: proc_macro::TokenStream,
+    item: proc_macro::TokenStream,
+) -> proc_macro::TokenStream {
+    let mut item = parse_macro_input!(item as ItemTrait);
+
+    let client = generate_client(&mut item);
+
+    quote! {
+        #item
+
+        #client
+    }
+    .into()
+}
+
+/// A no-op attribute macro required due to technical limitations of Rust's macro system.
+#[proc_macro_attribute]
+pub fn endpoint(
+    _attr: proc_macro::TokenStream,
+    item: proc_macro::TokenStream,
+) -> proc_macro::TokenStream {
+    item
+}
+
+fn generate_client(trait_: &mut ItemTrait) -> TokenStream {
+    let vis = &trait_.vis;
+    let trait_name = &trait_.ident;
+    let type_name = Ident::new(&format!("{}Client", trait_name), trait_name.span());
+
+    let asyncness = match resolve_asyncness(trait_) {
+        Ok(asyncness) => asyncness,
+        Err(e) => return e.into_compile_error(),
+    };
+
+    let service_trait = match asyncness {
+        Asyncness::Sync => quote!(Service),
+        Asyncness::Async => quote!(AsyncService),
+    };
+
+    let impl_attrs = match asyncness {
+        Asyncness::Sync => quote!(),
+        Asyncness::Async => quote!(#[conjure_http::private::async_trait]),
+    };
+
+    let where_ = match asyncness {
+        Asyncness::Sync => quote!(C: conjure_http::client::Client),
+        Asyncness::Async => quote! {
+            C: conjure_http::client::AsyncClient + Sync + Send,
+            C::ResponseBody: 'static + Send,
+        },
+    };
+
+    let methods = trait_
+        .items
+        .iter_mut()
+        .filter_map(|item| match item {
+            TraitItem::Fn(meth) => Some(meth),
+            _ => None,
+        })
+        .map(|m| generate_client_method(trait_name, asyncness, m));
+
+    quote! {
+        #vis struct #type_name<C> {
+            client: C,
+        }
+
+        impl<C> conjure_http::client::#service_trait<C> for #type_name<C> {
+            fn new(client: C) -> Self {
+                #type_name { client }
+            }
+        }
+
+        #impl_attrs
+        impl<C> #trait_name for #type_name<C>
+        where #where_
+        {
+            #(#methods)*
+        }
+    }
+}
+
+#[derive(Copy, Clone)]
+enum Asyncness {
+    Sync,
+    Async,
+}
+
+fn resolve_asyncness(trait_: &ItemTrait) -> Result<Asyncness, Error> {
+    let mut it = trait_.items.iter().filter_map(|t| match t {
+        TraitItem::Fn(f) => Some(f),
+        _ => None,
+    });
+
+    let Some(first) = it.next() else {
+        return Ok(Asyncness::Sync);
+    };
+
+    let is_async = first.sig.asyncness.is_some();
+
+    for f in it {
+        if f.sig.asyncness.is_some() != is_async {
```

Comment:
Do we want to enforce that all of the methods in a custom client are either async or blocking? I'm not sure if there is a technical reason that we would need something like this, but I don't think that java dialogue client annotations enforce this?

Reply:
Unfortunately it has to be all or none - the generated code looks something like

```rust
pub struct MyServiceClient<T> {
    client: T,
}


impl<T> MyService for MyServiceClient<T>
where
    T: Client, // or T: AsyncClient
{
    ...
}
```

We need to bound the inner raw client type on either the async client trait or the blocking client trait, which means that all of the methods need to be the same asyncness.

In practice it shouldn't matter, since the asyncness of the generated client will depend on the asyncness of the server using it which is typically also an all-or-none thing.

Reply:
The Java logic handles this with the inner client always returning a `ListenableFuture` and the generated code either returns that or `.getUnchecked()` it, but asyncness is a more "viral" feature in Rust than our async Java clients.